dissect/evidence/asdf/asdf.py

```diff
@@ -472,26 +584,101 @@ def _read(self, offset, length):
         return b"".join(r)
 
 
-class Crc32Stream(SubStreamBase):
-    """Compute a CRC32 over all written data.
-
-    This assumes that all data is written as a continuous stream.
+def scrape_blocks(fh: BinaryIO, buffer_size: int = io.DEFAULT_BUFFER_SIZE) -> Iterator[cstruct.Instance, int]:
+    """Scrape for block headers in ``fh`` and yield parsed block headers and their offset.
 
     Args:
-        fh: The file-like object to wrap.
+        fh: The file-like object to scrape for block headers.
+        buffer_size: The buffer size to use when scraping.
     """
+    # If BLOCK_MAGIC is not found in the buffer (4 bytes), it's possible that part of it (up to 3 bytes) is in there
+    # Keep an overlap buffer of 3 bytes that we prepend to the current buffer so we can also find these partial needles
+    overlap_len = len(BLOCK_MAGIC) - 1
+    overlap = b"\x00" * overlap_len
+    while True:
+        pos = fh.tell()
+        buf = fh.read(buffer_size)
+        if not buf:
+            break
+
+        data = overlap + buf
+        needle_pos = -1
+        while True:
+            needle_pos = data.find(BLOCK_MAGIC, needle_pos + 1)
+            if needle_pos == -1:
+                break
+
+            offset = pos + needle_pos - overlap_len
+
+            fh.seek(offset)
+            block_buf = fh.read(len(c_asdf.block))
+
+            # Some sanity checks that this is actually a block header
+            if block_buf[4] not in c_asdf.BLOCK_FLAG.reverse:
+                continue
+
+            if block_buf[6:8] != b"\x00\x00":
+                continue
+
+            block_entry = c_asdf.block(block_buf)
+            yield block_entry, offset
 
-    def __init__(self, fh):
-        super().__init__(fh)
-        self.crc = 0
+        # Keep the last 3 bytes as overlap
+        overlap = data[-overlap_len:]
+        # Consumer may seek the fh, so seek it back to where we were
+        fh.seek(pos + len(buf))
```

Comment:
I don't think so, `BLOCK_MAGIC` would have to be within `buf` and we're seeking past the size of `buf`. The unexpected side-effect is much worse since it may result in missing blocks or too large blocks. 

In some cases of scraping, you'd actually want the consumer to seek the `fh` since it might allow you to skip large portions of data. However in this case we don't know how large a block is in the `fh` and we actually determine it by looking for the next block, so we have to continue from exactly where we left of in order to correctly find the next block.